src/System.Diagnostics.TraceSource/src/System/Diagnostics/SwitchAttribute.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Constructor |
+                    AttributeTargets.Event | AttributeTargets.Method | AttributeTargets.Property)]
+    public sealed class SwitchAttribute : Attribute
+    {
+        private Type _type;
+        private string _name;
+        
+        public SwitchAttribute(string switchName, Type switchType)
+        {
+            SwitchName = switchName;
+            SwitchType = switchType;
+        }
+
+        public string SwitchName
+        {
+            get { return _name; }
+            set
+            {
+                if (value == null)
+                    throw new ArgumentNullException(nameof(value));
+                if (value.Length == 0)
+                    throw new ArgumentException(SR.Format(SR.InvalidNullEmptyArgument, nameof(value)), nameof(value));
+
+                _name = value;
+            }
+        }
+
+        public Type SwitchType
+        {
+            get { return _type; }
+            set
+            {
+                if (value == null)
+                    throw new ArgumentNullException(nameof(value));
+                _type = value;
+            }
+        }
+
+        public string SwitchDescription { get; set; }
+
+        public static SwitchAttribute[] GetAll(Assembly assembly)
+        {
+            if (assembly == null)
+                throw new ArgumentNullException(nameof(assembly));
+
+            List<object> switchAttribs = new List<object>();
```

Comment:
Nit: could this be a ```List<SwitchAttribute>```?  Then the allocating-and-copying-to-a-```SwitchAttribute[]``` at the end of the method could just be a ToArray call.